nixos/modules/virtualisation/google-compute-image.nix

```diff
@@ -150,20 +150,25 @@ in
                 fi
           fi
 
-          echo "obtaining SSH private host key..."
-          $wget -O /root/ssh_host_ecdsa_key  http://metadata/0.1/meta-data/attributes/ssh_host_ecdsa_key
-          if [ $? -eq 0 -a -e /root/ssh_host_ecdsa_key ]; then
-              mv -f /root/ssh_host_ecdsa_key /etc/ssh/ssh_host_ecdsa_key
-              echo "downloaded ssh_host_ecdsa_key"
-              chmod 600 /etc/ssh/ssh_host_ecdsa_key
-          fi
-
-          echo "obtaining SSH public host key..."
-          $wget -O /root/ssh_host_ecdsa_key.pub http://metadata/0.1/meta-data/attributes/ssh_host_ecdsa_key_pub
-          if [ $? -eq 0 -a -e /root/ssh_host_ecdsa_key.pub ]; then
-              mv -f /root/ssh_host_ecdsa_key.pub /etc/ssh/ssh_host_ecdsa_key.pub
-              echo "downloaded ssh_host_ecdsa_key.pub"
-              chmod 644 /etc/ssh/ssh_host_ecdsa_key.pub
+          countKeys=0
+          ${flip concatMapStrings config.services.openssh.hostKeys (k : ''
+            echo "trying to obtain SSH private host key ${baseNameOf k.path}"
+            ${wget} -O /root/${baseNameOf k.path} http://metadata/0.1/meta-data/attributes/${baseNameOf k.path} && :
+            if [ $? -eq 0 -a -e /root/${baseNameOf k.path} ]; then
+                countKeys=$((countKeys+1))
+                mv -f /root/${baseNameOf k.path} ${k.path}
+                echo "downloaded ${k.path}"
+                chmod 600 ${k.path}
```

Comment:
It would be better to set `umask 077` for the `wget` call so that the key is never world-readable. Theoretical, I guess.